core/src/main/scala/tech/ant8e/uuid4cats/UUIDUtil.scala

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2023 Antoine Comte
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tech.ant8e.uuid4cats
+
+import java.time.Instant
+import java.util.UUID
+
+object UUIDUtil {
+
+  /** Extract a timestamp from an UUID when it exists */
+  def extractTimestamp(uuid: UUID): Option[Instant] = {
+    import Extractors.*
+    (uuid match {
+      case UUIDv1(ts) => Some(ts)
+      case UUIDv6(ts) => Some(ts)
+      case UUIDv7(ts) => Some(ts)
+      case _          => None
+    }).map(Instant.ofEpochMilli)
+  }
+
+  private object Extractors {
+    object UUIDv1 {
+      def unapply(uuid: UUID): Option[Long] =
+        if (uuid.version == 1) {
```

Comment:
This could be an `Option.when(uuid.version == 1)` if one wanted to avoid the `else` block.